### ISSUE TITLE
int.__truediv__() accepted by Protocol that effectively requires that an int is returned

### ISSUE THREAD

**Issue by Joshix-1** [python/mypy#17110]
**Bug Report**

Mypy allows type-unsafe code.

**To Reproduce**

https://mypy-play.net/?mypy=1.0.0&python=3.10&gist=ab7449de421b39046dadf9c022d82e77

```py
from abc import abstractmethod
from typing import Protocol, TypeVar

from typing_extensions import reveal_type

T = TypeVar("T")


class SupportsSelfDivision(Protocol):

    @abstractmethod
    def __truediv__(self: T, other: int) -> T: ...


a: SupportsSelfDivision = 0  # ok (why?)


SSD = TypeVar("SSD", bound=SupportsSelfDivision)


def slash_1_000(x: SSD) -> SSD:
    return x / 1_000


reveal_type(slash_1_000(1))  # int (it's float at runtime!)
reveal_type(1 / 1_000)  # float

```

**Expected Behavior**

int should not satisfy SupportsSelfDivision. Runtime type should be the same as mypy reveals. Or there should be an error that int is not SupportsSelfDivision

**Actual Behavior**

int is SupportsSelfDivision. Mypy doesn't complain and says it's int even though it's float.

**Your Environment**

mypy Playground + local

- Mypy version used: 1.9.0
- Mypy command-line flags: -
- Mypy configuration options from `mypy.ini` (and other config files): -
- Python version used: 3.11.8

**Comment by ilevkivskyi** [comment]
This now gives an error as expected.
